tools/pythonpkg/src/pyconnection.cpp

```diff
@@ -137,6 +137,7 @@ DuckDBPyConnection *DuckDBPyConnection::Execute(const string &query, py::object
 	result = nullptr;
 	unique_ptr<PreparedStatement> prep;
 	{
+		const std::lock_guard<std::mutex> lock(py_connection_lock);
```

Comment:
Won't this lead to a race condition if T1 fires a query, then T2 tries to fire a query?

T1: lock(py_connection_lock) -> unlock(GIL) -> **lock(GIL)**
T2: lock(GIL) -> **lock(py_connection_lock)**

Now T1 holds `py_connection_lock` while waiting for the `GIL`, while T2 holds the `GIL` while waiting for the `py_connection_lock`, so you end up with a lock cycle.

I think for this to work correctly the thread first needs to release the `GIL` before attempting to grab the `py_connection_lock`, so you would end up with something like this:

```cpp
unique_lock<std::mutex> lock(py_connection_lock);
{
   py::gil_scoped_release release;
   lock.lock();
}
```

Reply:
Yes, that sounds about right.

However, I'm not sure where to introduce the lock then. 
The GIL is currently released in ```CompletePendingQuery()``` and acquired back right after to allow the users to cancel query results., so we can't really hold the connection lock, while releasing GIL during the whole query execution.
Any suggestions?

Reply:
My example only releases the GIL while obtaining the lock, which should be sufficient, no?